lib/authentication/auth_okta.js

```diff
@@ -49,46 +51,61 @@ function AuthOkta(password, region, account, clientType, clientVersion, httpClie
   * @returns {null}
   */
   this.authenticate = async function (authenticator, serviceName, account, username) {
-    let ssoUrl;
-    let tokenUrl;
-    await step1(authenticator, serviceName, account, username).then((response) => {
-      const responseData = response['data'];
-      const success = responseData['success'];
-      const errorCode = responseData['code'];
-      const errorMessage = responseData['message'];
-
-      if (typeof success === 'undefined' || errorCode === 'undefined' || errorMessage === 'undefined') {
-        throw new Error('Unable to use provided Okta address as an authenticator. Is the authenticator URL correct?');
-      }
+    const response = await step1(authenticator, serviceName, account, username);
+    const responseData = response['data'];
+    const success = responseData['success'];
+    const errorCode = responseData['code'];
+    const errorMessage = responseData['message'];
+    user = username;
 
-      if (success !== true) {
-        throw new Error(`Unable to use provided Okta address as an authenticator. Error code: ${errorCode}, error message: ${errorMessage}`);
-      }
+    if (typeof success === 'undefined' || errorCode === 'undefined' || errorMessage === 'undefined') {
+      throw new Error('Unable to use provided Okta address as an authenticator. Is the authenticator URL correct?');
+    }
 
-      ssoUrl = responseData['data']['ssoUrl'];
-      tokenUrl = responseData['data']['tokenUrl'];
-    });
+    if (success !== true) {
+      throw new Error(`Unable to use provided Okta address as an authenticator. Error code: ${errorCode}, error message: ${errorMessage}`);
+    }
 
-    step2(authenticator, ssoUrl, tokenUrl);
+    ssoUrl = responseData['data']['ssoUrl'];
+    tokenUrl = responseData['data']['tokenUrl'];
 
-    let oneTimeToken;
-    await step3(tokenUrl, username, password).then((response) => {
-      const data = response['data'];
+    step2(authenticator, ssoUrl, tokenUrl);
+    
+    const responseHtml = await step4( await step3(tokenUrl, username, password), ssoUrl);
 
-      if (data['sessionToken']) {
-        oneTimeToken = data['sessionToken'];
-      } else {
-        oneTimeToken = data['cookieToken'];
-      }
-    });
+    step5(responseHtml);
+  };
 
+  this.reauthenticate = async function (body, retryOption) {
+    const maxRetryTimeout = connectionConfig.getRetryTimeout();
+    const maxRetry = connectionConfig.getRetrySfMaxLoginRetries();
+    const remainingTimeout = maxRetryTimeout !== 0 ? (maxRetryTimeout - retryOption.totalElapsedTime) * 1000 : 0;
+    const startTime = Date.now();
     let responseHtml;
-    await step4(oneTimeToken, ssoUrl).then((response) => {
-      responseHtml = response['data'];
-    });
 
+    while (remainingTimeout === 0 || Date.now() < startTime + remainingTimeout) {
```

Comment:
In what cases remainingTimeout could be 0? This seems counterintuitive for me, because if the remainingTimeout reaches 0 shouldn't we stop retrying?

Reply:
Also, shouldn't we have here also a condition that checks numRetries against maxRetryCount?

Reply:
Hello @sfc-gh-dheyman-1, Thank you for checking. You are right. We do not need remainingTimeout condition here because we already declared the maxRetryTimeout. As a result, I changed the condition to 
`while (maxRetryTimeout=== 0 || Date.now() < startTime + remainingTimeout)`
Also, I added the maxRetryCount condition.